ocaml/testsuite/tests/pattern-guards/test.ml

```diff
@@ -688,3 +688,52 @@ check_push_defaults g ~s:"hmm" "name";;
 [%%expect{|
 - : string = "name"
 |}];;
+
+(* Test that type information introduced in a guard is available in the rhs. *)
+
+module Foo = struct
+  type t = A | B
+
+  let is_a = function
+    | A -> true
+    | B -> false
+end
+[%%expect{|
+module Foo : sig type t = A | B val is_a : t -> bool end
+|}];;
+
+let disambiguation_impossible = function
+  | None -> "None"
+  | Some y ->
+      match y with
+      | A -> "A"
+      | B -> "B"
+;;
+[%%expect{|
+Line 5, characters 8-9:
+5 |       | A -> "A"
+            ^
+Error: Unbound constructor A
+|}];;
+
+let disambiguation_from_guard = function
+  | None -> "None"
+  | Some y when Foo.is_a y -> "A"
+  | Some y ->
+      match y with
+      | A -> assert false
+      | B -> "B"
+;;
+[%%expect{|
+val disambiguation_from_guard : Foo.t option -> string = <fun>
+|}, Principal{|
+Line 6, characters 8-9:
+6 |       | A -> assert false
+            ^
+Warning 18 [not-principal]: this type-based constructor disambiguation is not principal.
+Line 7, characters 8-9:
+7 |       | B -> "B"
+            ^
+Warning 18 [not-principal]: this type-based constructor disambiguation is not principal.
+val disambiguation_from_guard : Foo.t option -> string = <fun>
+|}];;
```

Comment:
I think we can get rid of this principality warning if we move the match on `y` to be inside the case with the `when` guard, as opposed to an entirely unrelated case.

Reply:
I think moving it inside the associated case actually tests the thing we want, but even prior to this change, it would issue a principality warning.